stress: failed test in cockroach/kv/kv.test: TestClientNotReady

Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/6a7f101fdcc8adda1e84d85b5546396964b622fa

Stress build found a failed test:

```
=== RUN   TestClientNotReady
I160312 05:14:07.366731 /go/src/google.golang.org/grpc/server.go:454  grpc: Server.processUnaryRPC failed to write status stream error: code = 1 desc = "context canceled"
--- FAIL: TestClientNotReady (0.02s)
	send_test.go:286: kv/send_test.go:265: unexpected error when client was closed: failed to send RPC: too many errors encountered (1 of 1 total): rpc error: code = 2 desc = "grpc: the client connection is closing\n```

Run Details:

```
0 runs so far, 0 failures, over 5s
0 runs so far, 0 failures, over 10s
1 runs so far, 0 failures, over 15s
8 runs so far, 0 failures, over 20s
8 runs so far, 0 failures, over 25s
12 runs so far, 0 failures, over 30s
16 runs so far, 0 failures, over 35s
16 runs so far, 0 failures, over 40s
20 runs so far, 0 failures, over 45s
23 runs so far, 0 failures, over 50s
24 runs so far, 0 failures, over 55s
26 runs so far, 0 failures, over 1m0s
32 runs so far, 0 failures, over 1m5s
33 runs so far, 0 failures, over 1m10s
35 runs so far, 0 failures, over 1m15s
37 runs so far, 0 failures, over 1m20s
41 runs so far, 0 failures, over 1m25s
42 runs so far, 0 failures, over 1m30s
46 runs so far, 0 failures, over 1m35s
49 runs so far, 0 failures, over 1m40s
51 runs so far, 0 failures, over 1m45s
54 runs so far, 0 failures, over 1m50s
57 runs so far, 0 failures, over 1m55s
60 runs so far, 0 failures, over 2m0s
63 runs so far, 0 failures, over 2m5s
65 runs so far, 0 failures, over 2m10s
67 runs so far, 0 failures, over 2m15s
70 runs so far, 0 failures, over 2m20s
73 runs so far, 0 failures, over 2m25s
75 runs so far, 0 failures, over 2m30s
78 runs so far, 0 failures, over 2m35s
82 runs so far, 0 failures, over 2m40s
84 runs so far, 0 failures, over 2m45s
87 runs so far, 0 failures, over 2m50s
90 runs so far, 0 failures, over 2m55s
92 runs so far, 0 failures, over 3m0s
96 runs so far, 0 failures, over 3m5s
98 runs so far, 0 failures, over 3m10s
100 runs so far, 0 failures, over 3m15s
105 runs so far, 0 failures, over 3m20s
107 runs so far, 0 failures, over 3m25s
107 runs so far, 0 failures, over 3m30s
112 runs so far, 0 failures, over 3m35s
115 runs so far, 0 failures, over 3m40s
115 runs so far, 0 failures, over 3m45s
120 runs so far, 0 failures, over 3m50s
121 runs so far, 0 failures, over 3m55s
124 runs so far, 0 failures, over 4m0s
128 runs so far, 0 failures, over 4m5s
129 runs so far, 0 failures, over 4m10s
132 runs so far, 0 failures, over 4m15s
136 runs so far, 0 failures, over 4m20s
137 runs so far, 0 failures, over 4m25s
139 runs so far, 0 failures, over 4m30s
143 runs so far, 0 failures, over 4m35s
145 runs so far, 0 failures, over 4m40s
147 runs so far, 0 failures, over 4m45s
151 runs so far, 0 failures, over 4m50s
153 runs so far, 0 failures, over 4m55s
156 runs so far, 0 failures, over 5m0s
158 runs so far, 0 failures, over 5m5s
161 runs so far, 0 failures, over 5m10s
165 runs so far, 0 failures, over 5m15s
166 runs so far, 0 failures, over 5m20s
168 runs so far, 0 failures, over 5m25s
173 runs so far, 0 failures, over 5m30s
175 runs so far, 0 failures, over 5m35s
177 runs so far, 0 failures, over 5m40s
181 runs so far, 0 failures, over 5m45s
183 runs so far, 0 failures, over 5m50s
186 runs so far, 0 failures, over 5m55s
189 runs so far, 0 failures, over 6m0s
191 runs so far, 0 failures, over 6m5s
193 runs so far, 0 failures, over 6m10s
197 runs so far, 0 failures, over 6m15s
199 runs so far, 0 failures, over 6m20s
201 runs so far, 0 failures, over 6m25s
204 runs so far, 0 failures, over 6m30s
208 runs so far, 0 failures, over 6m35s
209 runs so far, 0 failures, over 6m40s
214 runs so far, 0 failures, over 6m45s
215 runs so far, 0 failures, over 6m50s
217 runs so far, 0 failures, over 6m55s
221 runs so far, 0 failures, over 7m0s
223 runs completed, 1 failures, over 7m4s
FAIL
```
Please assign, take a look and update the issue accordingly.

--- a/kv/send_test.go
+++ b/kv/send_test.go
@@ -226,11 +226,13 @@ func TestClientNotReady(t *testing.T) {
 	stopper := stop.NewStopper()
 	defer stopper.Stop()
 
-	nodeContext := newNodeTestContext(nil, stopper)
-
-	// Construct a server that listens but doesn't do anything.
-	s, ln := newTestServer(t, nodeContext)
-	roachpb.RegisterInternalServer(s, Node(50*time.Millisecond))
+	// Construct a server that listens but doesn't do anything. Notice that we
+	// never start accepting connections on the listener.
+	ln, err := net.Listen(util.TestAddr.Network(), util.TestAddr.String())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer ln.Close()
 
 	sp := tracing.NewTracer().StartSpan("node test")
 	defer sp.Finish()
@@ -243,6 +245,7 @@ func TestClientNotReady(t *testing.T) {
 	}
 
 	// Send RPC to an address where no server is running.
+	nodeContext := newNodeTestContext(nil, stopper)
 	if _, err := sendBatch(opts, []net.Addr{ln.Addr()}, nodeContext); err != nil {
 		retryErr, ok := err.(retry.Retryable)
 		if !ok {
@@ -255,13 +258,30 @@ func TestClientNotReady(t *testing.T) {
 		t.Fatalf("Unexpected success")
 	}
 
-	// Send the RPC again with no timeout.
+	// Send the RPC again with no timeout. We create a new node context to ensure
+	// there is a new connection.
+	nodeContext = newNodeTestContext(nil, stopper)
 	opts.SendNextTimeout = 0
 	opts.Timeout = 0
 	c := make(chan error)
+	sent := make(chan struct{})
+
+	// Start a goroutine to accept the connection from the client. We'll close
+	// the sent channel after receiving the connection, thus ensuring that the
+	// RPC was sent before we closed the connection. We intentionally do not
+	// close the server connection as doing so triggers other gRPC code paths.
 	go func() {
-		if _, err := sendBatch(opts, []net.Addr{ln.Addr()}, nodeContext); !testutils.IsError(err, "(failed as client connection was closed|transport is closing)") {
-			c <- util.Errorf("unexpected error when client was closed: %v", err)
+		_, err := ln.Accept()
+		if err != nil {
+			c <- err
+		} else {
+			close(sent)
+		}
+	}()
+	go func() {
+		_, err := sendBatch(opts, []net.Addr{ln.Addr()}, nodeContext)
+		if !testutils.IsError(err, "failed as client connection was closed") {
+			c <- util.Errorf("unexpected error: %v", err)
 		}
 		close(c)
 	}()
@@ -269,7 +289,7 @@ func TestClientNotReady(t *testing.T) {
 	select {
 	case err := <-c:
 		t.Fatalf("Unexpected end of rpc call: %v", err)
-	case <-time.After(10 * time.Millisecond):
+	case <-sent:
 	}
 
 	// Grab the client for our invalid address and close it. This will cause the

